### PR TITLE
Update tutorial CI workflow to ignore NotImplementedError exceptions

### PR DESCRIPTION
**Pull request by mwaskom**
This change enhances the notebook processing workflow to ignore `NotImplementedError` exceptions when running through the instructor version of the notebook.

This change should allow us to avoid commenting-out code that depends on completed exercises. This is better for a few reasons:

- Students will encounter the `NotImplementedError` as they step through the tutorial, meaning that it will always be clear when they need to do something.
- Uncommenting code is annoying and prone to error, especially in Python where indentation level is semantic.
- The CI workflow can now verify that notebooks are written such that the only error a student should see with an incomplete exercise (note: not a "complete but wrong" exercise) is `NotImplementedError`. This is part of our style guide, but the CI workflow could not actually verify it.
- Avoiding the need to comment out code in exercises drastically simplifies the problem of doing exercise verification.

Almost all of the code in the diff is copied (with the license included) from the [nbclient](https://github.com/jupyter/nbclient) library. Unfortunately, the way the cell execution method is written, it calls out to a private method to do error handling. So our custom subclass patches both the public cell execution method and the private error handling method.

This basic approach could be extended to do other custom NMA things. For example, we could have the CI workflow be responsible for enabling XKCD mode on the solution cells.

### CODE REVIEW

**Comment by mwaskom** [comment]
Setting to draft mode so I can test it with a few notebooks.

File: ci/process_notebooks.py [hunk 1]
```diff
@@ -37,6 +42,167 @@
 )
 
 
+class NMAPreprocessor(ExecutePreprocessor):
+    """
+    Custom subclass of the ExecutePreprocessor for NMA tutorials.
```
Comment (patrickmineault) [inline ci/process_notebooks.py:47]: Can you add a link to the source of ExecutePreprocessor?

**Review by patrickmineault — CHANGES_REQUESTED** [review]
This seems reasonable. It's not the cleanest because of the duplication but less risky if there are breaking upstream changes to private functions.

File: ci/process_notebooks.py (hunk 1 above)
Comment (mwaskom) [reply ci/process_notebooks.py:178]: Done

**Review by patrickmineault — APPROVED** [review]
LGTM

**Pull request merged**